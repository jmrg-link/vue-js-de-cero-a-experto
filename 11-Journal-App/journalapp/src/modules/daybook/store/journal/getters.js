export const getEntriesByTerm = (/*{}*/) =>{

}
export const getEntryById = (/*{}*/) =>{

}

