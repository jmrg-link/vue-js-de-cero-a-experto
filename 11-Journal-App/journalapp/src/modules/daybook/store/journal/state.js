export default () =>({
    isLoading: true,
    entries: [
        {
            id: new Date().getTime(),
            date: new Date().toDateString(),
            text: 'Lorem ipsum dolor sit amet consectetur adipiscing, elit aenean vestibulum et conubia nascetur, purus mi volutpat fames a. ',
            picture: null, // https:// uri
        },
        {
            id: new Date().getTime() * 2,
            date: new Date().toDateString(),
            text: 'Lobortis tincidunt ac nam magnis tristique natoque suscipit elementum montes sociis bibendum, orci nisl mauris senectus interdum metus torquent luctus parturient aliquam quisque, mus etiam laoreet augue viverra felis dapibus volutpat tortor duis. Mi cum ridiculus magnis semper massa senectus in accumsan iaculis suspendisse eleifend blandit, non phasellus himenaeos fringilla convallis venenatis platea pulvinar tellus pellentesque.',
            picture: null, // https:// uri
        },{
            id: new Date().getTime() * 2.5,
            date: new Date().toDateString(),
            text: 'Ut varius lobortis posuere ultricies faucibus accumsan nec in mattis, eget augue torquent quam molestie feugiat nisl mollis',
            picture: null, // https:// uri
        },


    ]
})