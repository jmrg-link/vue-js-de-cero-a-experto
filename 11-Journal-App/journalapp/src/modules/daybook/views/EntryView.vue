<template>
<div class="entry-title d-flex justify-content-between p-2 ">
  <div>
    <span class="text-success fs-1 fw-bold">15 </span>
    <span class="mx-1 ds-3">Julio</span>
    <span class="mx-1 fs-4 fw-light">2021, jueves</span>
  </div>
  <div>
    <button class="btn btn-danger mx-2">
      Borrar
      <i class="fa fa-trash-alt"></i>
    </button>

    <button class="btn btn-warning mx-2">
      Subir foto
      <i class="fa fa-upload"></i>
    </button>
  </div>
</div>
  <hr>
  <div class="d-flex flex-column px-3 h-75">
    <textarea
    placeholder=" Â¿Que ha sucedido hoy? "
    ></textarea>
  </div>
  <Fab
    icon="fa-save"
  />
  <img
      src="https://images.unsplash.com/photo-1533450718592-29d45635f0a9?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1470&q=80"
      alt="entry-picture"
      class="img-thumbnail"
  >
</template>
<script>
import { defineAsyncComponent } from "vue";

export default {
  components:{
    Fab: defineAsyncComponent(() => import('../components/Fab.vue'))
  }
}
</script>
<style lang="scss" scoped>
textarea{
  font-size: 20px;
  border: none;
  height: 100%;
  &:focus{
    outline: none;
  }
}
img {
  width: 200px;
  position: fixed;
  right: 20px;
  bottom: 150px;
  box-shadow: 0 5 10 rgba($color:#000000, $alpha: 0.2);
;
}

</style>