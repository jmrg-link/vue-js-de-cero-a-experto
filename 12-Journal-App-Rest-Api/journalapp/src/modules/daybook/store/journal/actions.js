import journalApi from "@/api/journalApi";

export const loadEntries = async ({ commit }) => {
    const { data } = await journalApi.get('/entries.json')
    const entries = []
    for( let id of Object.keys(data) ){
        entries.push({
            id,
            ...data[id]
        })
    }
    //console.log(entries)
    commit('setEntries', entries)
}
export const updateEntry = async ({ commit }, entry) => {
    const { date , picture , text } = entry
    const dataToSave = {date , picture , text }
    const resp = await journalApi.put( `/entries/${entry.id}.json`,dataToSave)
    console.log(resp)

    commit('updateEntry', { ...entry })
    //console.log(entry,'actions')
}
export const createEntry = async (/*{ commit }*/) => {

}
