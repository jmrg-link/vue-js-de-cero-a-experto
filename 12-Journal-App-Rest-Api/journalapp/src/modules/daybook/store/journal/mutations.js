

export const setEntries = (state,entries) => {
    state.entries = [...state.entries, ...entries]
    state.isLoading = false
}
export const updateEntry = ( state , entry ) => {
    const index = state.entries.map(e => e.id).indexOf(entry.id)
    state.entries[index] = entry

}
export const addEntry = (/*{ state }*/) => {

}
