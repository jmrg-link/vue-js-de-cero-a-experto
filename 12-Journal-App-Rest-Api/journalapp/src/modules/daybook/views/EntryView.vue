<template>
  <template v-if="entry">
    <div class="entry-title d-flex justify-content-between p-2 ">
      <div>
        <span class="text-success fs-1 fw-bold">{{ day }}</span>
        <span class="mx-1 ds-3">{{ month }}</span>
        <span class="mx-1 fs-4 fw-light">{{ yearDay }}</span>
      </div>
      <div>
        <button class="btn btn-danger mx-2">
          Borrar
          <i class="fa fa-trash-alt"></i>
        </button>

        <button class="btn btn-warning mx-2">
          Subir foto
          <i class="fa fa-upload"></i>
        </button>
      </div>
    </div>
    <hr>
    <div class="d-flex flex-column px-3 h-75">
    <textarea
        v-model="entry.text"
        placeholder=" Â¿Que ha sucedido hoy? "
    ></textarea>
    </div>

    <img
        src="https://images.unsplash.com/photo-1533450718592-29d45635f0a9?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1470&q=80"
        alt="entry-picture"
        class="img-thumbnail"
    >
  </template>

  <Fab icon="fa-save"/>

</template>
<script>
import { defineAsyncComponent } from "vue";
import { mapGetters } from 'vuex' //computed
import getDayMonthYear from '../helpers/getDayMonthYear'

export default {
  props:{
    id:{
      type:String,
      required: true
    }
  },

  components:{
    Fab: defineAsyncComponent(() => import('../components/Fab.vue'))
  },

  methods: {
    loadEntry(){
      const entry = this.getEntryById(this.id)
      if(!entry) return this.$router.push({name:'no-entry'})
      this.entry = entry
      //console.log(entry)
    }
  },
  data(){
    return {
      entry: null
    }
  },
  computed: {
    ...mapGetters('journal',['getEntryById']),
    day(){
      const { day } = getDayMonthYear(this.entry.date)
      return day
    },
    month(){
      const { month } = getDayMonthYear(this.entry.date)
      return month
    },
    yearDay(){
      const { yearDay } = getDayMonthYear(this.entry.date)
      return yearDay
    },

  },

  created() {
    //console.log(this.$route.params.id)
    //console.log(this.id)
    this.loadEntry()
  },

  watch:{
    id(){
      this.loadEntry()
    }
  }
}
</script>
<style lang="scss" scoped>
textarea{
  font-size: 20px;
  border: none;
  height: 100%;
  &:focus{
    outline: none;
  }
}
img {
  width: 200px;
  position: fixed;
  right: 20px;
  bottom: 150px;
  box-shadow: 0 5 10 rgba($color:#000000, $alpha: 0.2);
;
}

</style>