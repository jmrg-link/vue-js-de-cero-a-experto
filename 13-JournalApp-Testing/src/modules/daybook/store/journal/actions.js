import journalApi from "@/api/journalApi";

export const loadEntries = async ({commit}) => {
    const {data} = await journalApi.get('/entries.json')
    if (!data) {
        commit('setEntries', [])
        return
    }

    const entries = []
    for (let id of Object.keys(data)) {
        entries.push({
            id,
            ...data[id]
        })
    }
    //console.log(entries)
    commit('setEntries', entries)
}
export const updateEntry = async ({commit}, entry) => { // entry debe ser un parametro
    const {date, picture, text} = entry
    const dataToSave = {date, picture, text}
    const resp = await journalApi.put(`/entries/${entry.id}.json`, dataToSave)
    dataToSave.id = entry.id // dentro de dataToSave tiene que ser igual que el contenido del object de entry.id
    // console.log(resp)
    // commit de una mutation -> updateEntry
    commit('updateEntry', {...dataToSave})
}

export const createEntry = async ({commit}, entry) => {
    const {date, picture, text} = entry
    const dataToSave = {date, picture, text}
    const {data} = await journalApi.post(`/entries.json`, dataToSave)
    dataToSave.id = data.name
    commit('addEntry', dataToSave)
    return data.name
}

export const deleteEntry = async ({commit}, id) => {
    await journalApi.delete(`/entries/${id}.json`)
    commit('deleteEntry', id)
    return id
}