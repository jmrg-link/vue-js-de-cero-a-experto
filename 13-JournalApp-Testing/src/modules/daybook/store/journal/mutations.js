

export const setEntries = (state,entries) => {
    state.entries = [...state.entries, ...entries]
    state.isLoading = false
}
export const updateEntry = ( state , entry ) => {
    const index = state.entries.map(e => e.id).indexOf(entry.id)
    state.entries[index] = entry

}
export const addEntry = ( state , entry ) => {
    state.entries = [ entry , ...state.entries ]
}

export const deleteEntry = (state, id)=> {
    state.entries = state.entries.filter(e => e.id !== id)
}