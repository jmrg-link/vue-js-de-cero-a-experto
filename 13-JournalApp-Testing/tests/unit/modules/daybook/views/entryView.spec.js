import {shallowMount} from '@vue/test-utils'
import {createStore} from 'vuex'

import journal from '@/modules/daybook/store/journal'
import {journalState} from '../../../mock-data/test-journal-state'

import EntryView from '@/modules/daybook/views/EntryView.vue'
import swal from "sweetalert2";

const createVuexStore = (initialState) =>
    createStore({
        modules: {
            journal: {
                ...journal,
                state: {...initialState}
            }
        }
    })

jest.mock('sweetalert2', () => ({
    fire: jest.fn(),
    showLoading: jest.fn(),
    close: jest.fn()
}))


describe('Pruebas en el EntryView', () => {
    const store = createVuexStore(journalState)
    store.dispatch = jest.fn()
    const mockRouter = {
        push: jest.fn()
    }
    let wrapper
    beforeEach(() => {
        jest.clearAllMocks()
        wrapper = shallowMount(EntryView, {
            props: {
                id: '-MqUePtgfaib9tio3XUF'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [store]
            }
        })
    })

    test('debe de sacar al usuario si el :ID no existe', () => {
        const wrapper = shallowMount(EntryView, {
            props: {
                id: 'Este ID no existe en el STORE'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [store],
            }
        })
        expect(mockRouter.push).toHaveBeenCalledWith({name: 'no-entry'})
    })

    test('debe de mostrar la entrada correctamente', () => {
        expect(wrapper.html()).toMatchSnapshot()
        expect(mockRouter.push).not.toHaveBeenCalled()
    })

    test('debe borrar la entrada y salir', async (done) => {
        swal.fire.mockReturnValueOnce(Promise.resolve({isConfirmed: true}))
        wrapper.find('.btn-danger').trigger('click')
        expect(swal.fire).toHaveBeenCalledWith({
            title: 'Â¿Estas seguro/a que quieres borrar esta entrada?',
            text: 'Si deseas continuar perderas el contenido de la entrada.',
            showDenyButton: true,
            confirmButtonText: 'Si, Estoy seguro/a'
        })

        setTimeout(() => {
            expect(store.dispatch).toHaveBeenCalledWith("journal/deleteEntry", "-MqUePtgfaib9tio3XUF")
            expect(mockRouter.push).toHaveBeenCalled()
            done()
        }, 200)
    })
})