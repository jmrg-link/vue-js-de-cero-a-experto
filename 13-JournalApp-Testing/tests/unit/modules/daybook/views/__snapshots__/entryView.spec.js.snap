// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el EntryView debe de mostrar la entrada correctamente 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-1 fw-bold">9</span><span class="mx-1 ds-3">Diciembre</span><span class="mx-1 fs-4 fw-light">2021, Jueves</span></div>
  <div><input type="file" accept="image/png, image/jpeg, image/jpg" style="display: none;"><button class="btn btn-danger mx-2"> Borrar <i class="fa fa-trash-alt"></i></button><button class="btn btn-warning mx-2"> Subir foto <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder=" Â¿Que ha sucedido hoy? "></textarea></div>
<img src="https://res.cloudinary.com/dqhfairbx/image/upload/v1639063054/iou33dhvmulfbegxl1g7.jpg" alt="entry-picture" class="img-thumbnail">
<!--v-if-->
<fab-stub icon="fa-save"></fab-stub>
`;
